src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/UnpooledDataSource.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NET8_0_OR_GREATER
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX
+{
+    /// <summary>
+    /// A data source that always creates a connection from scratch.
+    /// </summary>
+    internal sealed class UnpooledDataSource : SqlDataSource
+    {
+        /// <summary>
+        /// Initializes a new instance of UnpooledDataSource.
+        /// </summary>
+        /// <param name="connectionString"></param>
+        /// <param name="credential"></param>
+        internal UnpooledDataSource(string connectionString, SqlCredential credential) : base(connectionString, credential)
+        {
+        }
+
+        /// <summary>
+        /// Creates and opens a new SqlConnector.
+        /// </summary>
+        /// <param name="async">Whether this method should be run asynchronously.</param>
+        /// <param name="cancellationToken">Cancels an outstanding asynchronous operation.</param>
+        /// <returns></returns>
+        internal override async ValueTask<SqlConnector> GetInternalConnection(bool async, CancellationToken cancellationToken)
+        {
+            var connection = new SqlConnector();
+            await connection.Open(async, cancellationToken).ConfigureAwait(false);
```

Comment:
In the synchronous case, the cancellation token needs to be checked after calling `Open`. If it's cancelled then you'll probably want to close the connection and throw an exception.

The same pattern applies in quite a few places - if the operation is cancelled before completion, we need to make sure the nested resources are disposed before throwing the cancellation exception. @benrr101 I've not mentioned this part of the cancellation checks before, it also needs to be considered when we plumb it in.

Reply:
Thanks, @edwardneal. I'm open to ideas if there's a better way to express this in the code, but if you trace the call to this method up to the "public" API surface (in SqlConnectionX), it will always pass down a `CancellationToken.None` for the sync case. My intention is to ignore cancellation tokens for the sync path. 

The public surface looks like this:
```c#
public override void Open() => Open(false, CancellationToken.None).GetAwaiter().GetResult();
```

For the async case, I'd consider this past the useful point of cancellation checking as all the work has already been done.

There is still work to be done to respect connect timeouts and I'll definitely add that in to this PR.

Reply:
I think part of this depends on how we want to handle those connection timeouts. Some of that work is yet to be done, but it seems to me that the simplest way would be to have a datastructure which converts a user-specified connection timeout into `CancellationToken` and which provides the number of seconds remaining until the timeout fully elapses (to allow the socket's connection timeout to be set appropriately in the synchronous case.) At that point, async cancellation support is just a matter of combining the two `CancellationToken` objects and passing it down the chain. In this scenario, `CancellationToken.None` would only be passed in the synchronous case with no connection timeouts.

Reply:
I agree. I believe the best approach will be to pass the timeout down to the network layer where this cancellation token orchestration can be owned and managed. I pushed changes that reflect this.

Reply:
I agree, with the caveat that the connection timeout is end-to-end, so if there's some non-network activity which takes a while (such as a wait on the connection pool's availability) then it needs to encompass that too. I'll experiment to see if there's a way to make the combination of timeout + `CancellationToken` generic enough to use for the command timeout too.

Reply:
I'll add a TODO and we can rework the API for specifying timeout as we get into the new connection pool implementation.